### ISSUE TITLE
Improve `fetch_remote_file()` to support post requests with other types of request body

### ISSUE THREAD

**Issue by euantorano** [mybb/mybb#3100]
The `fetch_remote_file()` function will currently accept a second parameter which is an array of post data. This data is converted into the form request format (`key=val`), though there are several web services that expect POST requests with JSON bodies and such. 

It would be good to extend this function to support other types of request. There are a few enhancements that would be very nice to include:

- If the post data parameter is a string, pass that as is as the post body. 
- Support other HTTP request methods such as `PUT`, `PATCH`, `DELETE`, etc. 
- Support setting the request content type. 

If there are any other enhancements that could be made, please list them here and I’ll see what I can do.

**Comment by effone** [comment]
Multiple request support?

**Comment by yuliu** [comment]
Reference a [community post](https://community.mybb.com/thread-224406.html) here for a discussion to disregard the letter case in header's `Location: `.

**Comment by euantorano** [comment]
HTTP headers are supposed to be case insensitive according to the specification if I remember correctly. If we’re doing case sensitive checks, that’s a bug for sure that should be fixed in feature.

**Comment by yuliu** [comment]
[RFC 7230](https://tools.ietf.org/html/rfc7230#section-3.2) says it's case-insensitive:

> 3.2.  Header Fields
>
>   Each header field consists of a case-insensitive field name followed
>   by a colon (":"), optional leading whitespace, the field value, and
>   optional trailing whitespace.

Seems these fields are case-sensitive: request method, URI (but not hostname) and some fields in Protocol Versioning.

**Comment by euantorano** [comment]
Yep, that’s what I thought. If we’re doing case sensitive checks, that’s 1000% a bug that needs fixed 😂

**Comment by Ben-MyBB** [comment]
Fixed in https://github.com/mybb/mybb/issues/3850
